src/token/token.service.ts

```diff
@@ -55,8 +117,164 @@ export class TokenService {
       const tokenResults = await axios.get(priceServiceUrl);
       return tokenResults.data;
     } catch (e) {
-      this.logger.warn(`Data request error: ${e}`);
+      this.logger.warn(`Data request error from price service: ${e}`);
+    }
+  }
+
+  async requestNearPrice() {
+    try {
+      const nearHelperUrl = configService.getHelperUrl();
+      const nearData = await axios.get(`${nearHelperUrl}/fiat`);
+      return nearData.data.near.usd;
+    } catch (e) {
+      this.logger.warn(`Data request error from helper: ${e}`);
+    }
+  }
+
+  async getDataFromPools() {
+    try {
+      const service = await initializeNearService();
+      const length = await service.viewFunction('get_number_of_pools');
+      const pages = Math.ceil(length / DEFAULT_PAGE_LIMIT);
+      const pools = await Promise.allSettled(
+        [...Array(pages)].map((_, i) =>
+          service.viewFunction('get_pools', {
+            from_index: i * DEFAULT_PAGE_LIMIT,
+            limit: DEFAULT_PAGE_LIMIT,
+          }),
+        ),
+      );
+      return pools
+        .filter(assertFulfilled)
+        .map(({ value }) => value)
+        .flat()
+        .map((pool) => {
+          return {
+            ...pool,
+            supplies: pool.amounts.reduce(
+              (
+                acc: { [tokenId: string]: string },
+                amount: string,
+                i: number,
+              ) => {
+                acc[pool.token_account_ids[i]] = amount;
+                return acc;
+              },
+              {},
+            ),
+          };
+        });
+    } catch (e) {
+      this.logger.warn(`Data request error from pools: ${e}`);
+      return [];
+    }
+  }
+
+  async calculatePrices(
+    poolsFromJumbo,
+    jumboPrice,
+    nearFiatPrice: string,
+  ): Promise<{
+    [key: string]: { price: string; volume: string; token: Token };
+  }> {
+    const jumboAddress = configService.getJumboTokenId();
+    const nearAddress = configService.getNearTokenId();
+
+    const newPrices = {};
+    for (const pool in poolsFromJumbo) {
```

Comment:
Suggestion: move the body of this loop into a separate private method and parallelize it with `Promise.all`.